---
title: TOX-induced exhaustion is required for the survival and persistence of antigen-specific CD8 T cells in tumors
#runtime: shiny_prerendered
output:
  html_document:
    code_folding: hide
    theme: space
    toc: yes
    toc_depth: 3
    toc_float: no
  BiocStyle::html_document2:
    code_folding: hide
    toc: yes
    toc_float: yes
  knitrBootstrap::bootstrap_document:
    highlight.chooser: yes
    theme.chooser: yes
  pdf_document:
    toc: yes
always_allow_html: yes
---


```{r setup, bootstrap.show.code = FALSE, results='hide', bootstrap.show.message=FALSE, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE}
knitr::opts_chunk$set(bootstrap.show.code = FALSE, message=FALSE, warning=FALSE)
suppressMessages(library(pheatmap))
suppressMessages(library(DESeq2))
suppressMessages(library(edgeR))
suppressMessages(library(ggplot2))
suppressMessages(library(ggrepel))
getOutputFormat <- function() {
  output <- rmarkdown:::parse_yaml_front_matter(
    readLines(knitr::current_input())
    )$output
  if (is.list(output)){
    return(names(output)[1])
  } else {
    return(output[1])
  }
}
if(getOutputFormat() == 'pdf_document') {
  knitr::opts_chunk$set(bootstrap.show.code = FALSE, message=FALSE, warning=FALSE, echo=FALSE, results='hide', plots='all')
}
tol12qualitative=c("#9e9ac8", "#6a51a3", "#fdbe85", "#fd8d3c", "#d94701", "#74c476", "#41ab5d", "#238b45", "#005a32", "#bdd7e7", "#6baed6", "#2171b5")
palette <- c("#d94701", "#6a51a3", "#2171b5", "#238b45")

```



```{r deseq2_figure2, message=FALSE, eval=TRUE, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, results='hide'}
counts <- read.table(file = "fig2.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "fig2.metadata.txt"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F,sep="\t")
decoder.data$condition <- factor(decoder.data$condition)
decoder.data$group.ID <-factor(decoder.data$condition)
decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]
#table(colnames(counts) == decoder.data$sample.ID)

sampleConditions <- factor(make.names(decoder.data$group.ID), levels=c("N", "KO","OT1","SV40","WT"))
deseq2.coldata <- data.frame(condition = sampleConditions, row.names = colnames(counts))
deseq2.dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, colData = deseq2.coldata, design = ~ condition)
isexpr <- rowSums(cpm(counts(deseq2.dds)) > 1) >= min(table(sampleConditions))
deseq2.dds <- deseq2.dds[isexpr,]
deseq2.dds <- DESeq(deseq2.dds)

deseq2_SV40_vs_OT1_sig <- as.data.frame(subset(results(deseq2.dds, contrast=c("condition","SV40","OT1"), alpha=0.05), padj < 0.05))
deseq2_SV40_vs_OT1_all <- as.data.frame(results(deseq2.dds, contrast=c("condition","SV40","OT1"), alpha=0.05))
```


# Figure 2

Figure 2e: Heatmap of RNA-Seq expression for genes differentially expressed between TAG and OT1 T cells (FDR<0.05). 

```{r figure2_e, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, height=10}
cpms <- cpm(counts[isexpr,], log=T)
my_colors = c("green", "black", "red")
my_colors = colorRampPalette(my_colors)(50)
pheatmap(cpms[row.names(deseq2_SV40_vs_OT1_sig),grep("SV40|OT1", colnames(cpms))], scale="row", fontsize=8, show_rownames=F, color = my_colors, main = "SV40 vs. OT1, padj < 0.05")
```


Figure 2f:  MA-plot of RNA-Seq data set; significantly differentially expressed genes are colored in red. 

```{r figure2_f, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE}
genes_to_highlight <- c("Tox", "Nfil3", "Tigit","Il10", "Pdcd1", "Havcr2", "Cd101", "Entpd1", "Klre1", "Id2", "Cd2", "Ccr7", "Tcf7", "Lef1", "Il7r")
res <- deseq2_SV40_vs_OT1_all
res <- res[with(res, order(log2FoldChange)),]
res <- res[which(!is.na(res$padj)),]
res$threshold  <- FALSE
res[res$padj < 0.05,]$threshold <- TRUE
res$gene <- row.names(res)
ggplot(data=res, aes(x=log2(baseMean), y=log2FoldChange, colour=threshold, label=gene)) + 
    geom_point(alpha=0.8, size=1, show.legend = FALSE) + 
    geom_hline(aes(yintercept = 0), colour = "red", size = 0.45) +
    ylim( c(-1, 1) * quantile(abs(res$log2FoldChange[is.finite(res$log2FoldChange)]), probs = 0.99) * 1.3) + 
    xlab("Mean expression") + 
    ylab("Log2 FC") + 
     theme_bw(base_size = 16) + theme(legend.position="bottom") +
     scale_colour_manual(values = c("gray80", "red3")) + ggtitle("SV40 vs. OT1")  + 
     geom_text_repel(data=data.frame(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE)  ), size=5, segment.color="white", color="black") + geom_point(data=subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE), shape=1, size=3, color=ifelse(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE)$log2FoldChange > 0, "red", "darkblue")) 
```


# Figure 3

```{r deseq2_figure3, message=FALSE, eval=TRUE, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, results='hide'}
counts <- read.table(file = "fig3.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "fig3.metadata.txt"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F,sep="\t")
decoder.data$condition <- factor(decoder.data$condition)
decoder.data$group.ID <-factor(decoder.data$condition)
decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]
# table(colnames(counts) == decoder.data$sample.ID)

cnts <- counts
conds <- factor(make.names(decoder.data$group.ID))

sampleConditions <-  factor(make.names(decoder.data$group.ID))
deseq2.coldata <- data.frame(condition = sampleConditions, row.names = colnames(counts))
deseq2.dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, colData = deseq2.coldata, design = ~ condition)
isexpr <- rowSums(cpm(counts(deseq2.dds)) > 1) >= min(table(sampleConditions))
deseq2.dds <- deseq2.dds[isexpr,]
deseq2.dds <- DESeq(deseq2.dds)

deseq2_TOX_vs_mGFP_sig <- as.data.frame(subset(results(deseq2.dds, contrast=c("condition","TOX","mGFP"), alpha=0.10), padj < 0.10))
deseq2_TOX_vs_mGFP_all <- as.data.frame(results(deseq2.dds, contrast=c("condition","TOX","mGFP"), alpha=0.10))
```

Figre 3d: Heatmap of RNA-Seq expression for genes differentially expressed between TOX-GFP and GFP-expressing TCR TAG (FDR<0.10)

```{r figure3_d, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, height=10}
cpms <- cpm(counts[isexpr,], log=T)
my_colors = c("green", "black", "red")
my_colors = colorRampPalette(my_colors)(50)
pheatmap(cpms[row.names(deseq2_TOX_vs_mGFP_sig),grep("TOX|mGFP", colnames(cpms))], scale="row", fontsize=8, show_rownames=F, color = my_colors, main = "Tox vs. mGFP, padj < 0.10")
```

Figre 3e: MA-plot of RNA-Seq data set; significantly differentially expressed genes are colored in red.

```{r figure3_e, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE}
genes_to_highlight <- c("Tox", "Cd244", "Eomes", "Havcr2", "Entpd1", "Id2", "Pdcd1", "Gzmb", "Cd38", "Cd160", "Cd101", "Fasl", "Ahr", "Nfil3", "Egr3", "Il5", "Tox", "Ahr","Nfil3", "Eomes", "Prdm1", "Id2", "Id3", "Lef1", "Tcf7","Egr3","Pdcd1", "Cd244", "Havcr2", "Entpd1","Tnfrsf9", "Tnfrsf8", "Klre1","Cd2", "Cd27", "Cxcr3", "Ccr7", "Gzma", "Gzmb","Gzmc",  "Il5", "Il10", "Lef1", "Tcf7", "Tox", "Entpd1", "Cd244", "Tigit", "Cd38", "Il7r", "Tcf7l2")
genes_to_highlight <- unique(genes_to_highlight)
res <- deseq2_TOX_vs_mGFP_all
res <- res[with(res, order(log2FoldChange)),]
res <- res[which(!is.na(res$padj)),]
res$threshold  <- FALSE
res[res$padj < 0.10,]$threshold <- TRUE
res$gene <- row.names(res)
ggplot(data=res, aes(x=log2(baseMean), y=log2FoldChange, colour=threshold, label=gene)) + 
    geom_point(alpha=0.8, size=1, show.legend = FALSE) + 
    geom_hline(aes(yintercept = 0), colour = "red", size = 0.45) +
    xlab("Mean expression") + 
    ylab("Log2 FC") + 
    ylim(-3,3) +
    theme_bw(base_size = 16) + theme(legend.position="bottom") +
    scale_colour_manual(values = c("gray80", "red3")) + ggtitle("Tox vs. GFP")  + 
    geom_text_repel(data=data.frame(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE)  ), size=5, segment.color="white", color="black") + geom_point(data=subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE), shape=1, size=3, color=ifelse(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE)$log2FoldChange > 0, "red", "darkblue")) 
```

Figre 3g: Heatmap of RNA-Seq expression for differentially expressed genes encoding transcription factors, inhibitory receptors, activation and signaling molecules, as well as cytokines and effector molecules in TOX-GFP overexpressing (TOX) or GFP-only expressing (GFP) TCR TAG. 

```{r figure3_g, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, height=10}
breaksList = seq(-1.5,1.5, by = .1)
color = colorRampPalette(c('#0C355E', 'white', '#9D2F2F'))(length(breaksList))
color = colorRampPalette(c('blue', 'white', 'red'))(length(breaksList))
cellwidth = 25
cellheight = 25
rlog_tran <- rlog(as.matrix(counts))
row.names(rlog_tran) <- row.names(counts)

tran_factors <- c("Tox", "Ahr","Nfil3", "Eomes", "Prdm1", "Id2", "Id3", "Lef1", "Tcf7","Egr3")
#tran_factors %in% row.names(deseq2_TOX_vs_mGFP_sig)
tran_factors_lbls <- tran_factors
tran_factors_lbls[which( tran_factors %in% row.names(deseq2_TOX_vs_mGFP_sig) == TRUE)] <- paste0(tran_factors[which( tran_factors %in% row.names(deseq2_TOX_vs_mGFP_sig) == TRUE)], "*")
pheatmap(rlog_tran[tran_factors,], scale="row",  cluster_rows = FALSE, cluster_cols = F, gaps_col = c(2),  main="Transcription factors", color = color, border_color = NA,  cellwidth = cellwidth, cellheight = cellheight,  breaks = breaksList)

selected_genes <- c("Pdcd1", "Cd244", "Havcr2", "Entpd1")
#selected_genes %in% row.names(deseq2_TOX_vs_mGFP_sig)
selected_genes_lbls <- selected_genes
pheatmap(rlog_tran[selected_genes,], scale="row", cluster_rows = FALSE, cluster_cols = F, gaps_col = c(2), main="Inhibitory Receptors", color = color, border_color = NA,  cellwidth = cellwidth, cellheight = cellheight,  breaks = breaksList)

selected_genes <- c("Tnfrsf9", "Tnfrsf8", "Klre1","Cd2", "Cd27", "Cxcr3", "Ccr7") 
#selected_genes %in% row.names(deseq2_TOX_vs_mGFP_sig)
selected_genes_lbls <- selected_genes
pheatmap(rlog_tran[selected_genes,], scale="row", labels_row=selected_genes_lbls, cluster_rows = FALSE, cluster_cols = F, gaps_col = c(2),color=color, main="Activation and Memory Markers", border_color = NA,  cellwidth = cellwidth, cellheight = cellheight,  breaks = breaksList)

selected_genes <- c( "Gzma", "Gzmb","Gzmc",  "Il5", "Il10")
#selected_genes %in% row.names(deseq2_TOX_vs_mGFP_sig)
selected_genes_lbls <- selected_genes
pheatmap(rlog_tran[selected_genes,], scale="row", cluster_rows = FALSE, cluster_cols = F, gaps_col = c(2), color=color, main="Cytokines and Effector Molecules", border_color = NA,  cellwidth = cellwidth, cellheight = cellheight, breaks = breaksList)

```


# Figure 5


```{r deseq2_figure5, message=FALSE, eval=TRUE, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, results='hide'}
counts <- read.table(file = "fig5.counts.txt", header = TRUE, check.names=FALSE, row.names=1)
decoderFile <- "fig5.metadata.txt"
decoder.data <- read.table(decoderFile,header=T,stringsAsFactors=F,sep="\t")
decoder.data$condition <- factor(decoder.data$condition)
decoder.data$group.ID <-factor(decoder.data$condition)
decoder.data <- decoder.data[decoder.data$sample.ID %in% colnames(counts),]
# table(colnames(counts) == decoder.data$sample.ID)

sampleConditions <- factor(make.names(decoder.data$group.ID), levels=c("N", "KO","OT1","SV40","WT"))
deseq2.coldata <- data.frame(condition = sampleConditions, row.names = colnames(counts))
deseq2.dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, colData = deseq2.coldata, design = ~ condition)
isexpr <- rowSums(cpm(counts(deseq2.dds)) > 1) >= min(table(sampleConditions))
deseq2.dds <- deseq2.dds[isexpr,]
deseq2.dds <- DESeq(deseq2.dds)

deseq2_WT_vs_KO_sig <- as.data.frame(subset(results(deseq2.dds, contrast=c("condition","KO","WT"), alpha=0.05), padj < 0.05))
deseq2_WT_vs_KO_all <- as.data.frame(results(deseq2.dds, contrast=c("condition","KO","WT"), alpha=0.05))
```


Figure 5a: Heatmap of RNA-Seq expression for genes differentially expressed between WT and KO cells (FDR<0.05). 

```{r figure5_a, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE, height=10}
cpms <- cpm(counts[isexpr,], log=T)
my_colors = c("green", "black", "red")
my_colors = colorRampPalette(my_colors)(50)
pheatmap(cpms[row.names(deseq2_WT_vs_KO_sig),grep("KO|WT", colnames(cpms))], scale="row", fontsize=8, show_rownames=F, color = my_colors, main = "KO vs. WT, padj < 0.05")
```



Figure 5b:  MA-plot of RNA-Seq data set; significantly differentially expressed genes are colored in red. 

```{r figure5_b, warning=FALSE, cache=TRUE, echo=FALSE, comment=FALSE}
genes_to_highlight <- c("Tox", "Il7r", "Lef1", "Tcf7", "Cd38", "Cd101", "Pdcd1", "Entpd1", "Tigit", "Prdm1", "Havcr2", "Nfil3")
res <- deseq2_WT_vs_KO_all
res <- res[with(res, order(log2FoldChange)),]
res <- res[which(!is.na(res$padj)),]
res$threshold  <- FALSE
res[res$padj < 0.05,]$threshold <- TRUE
res$gene <- row.names(res)
ggplot(data=res, aes(x=log2(baseMean), y=log2FoldChange, colour=threshold, label=gene)) + 
    geom_point(alpha=0.8, size=1, show.legend = FALSE) + 
    geom_hline(aes(yintercept = 0), colour = "red", size = 0.45) +
    ylim(-10, 10) +
    xlab("Mean expression") + 
    ylab("Log2 FC") + 
     theme_bw(base_size = 16) + theme(legend.position="bottom") +
     scale_colour_manual(values = c("gray80", "red3")) + ggtitle("WT vs. KO")  + 
     geom_text_repel(data=data.frame(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE || threshold == FALSE )  ), size=5, segment.color="white", color="black") + geom_point(data=subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE), shape=1, size=3, color=ifelse(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == TRUE)$log2FoldChange > 0, "red", "darkblue")) + geom_point(data=subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == FALSE), shape=1, size=3, color=ifelse(subset(res[genes_to_highlight[genes_to_highlight %in% row.names(res)],], threshold == FALSE)$log2FoldChange > 0, "black", "black"))
```

